pkg/policy/mapstate.go

```diff
@@ -276,6 +276,102 @@ func (ms *mapState) SubsetKeysWithSameID(key Key) iter.Seq2[Key, mapStateEntry]
 	}
 }
 
+// LPMAncestors iterates over broader or equal port/proto entries in the trie in LPM order,
+// with most specific match with the same ID as in 'key' being returned first.
+func (ms *mapState) LPMAncestors(key Key) iter.Seq2[Key, mapStateEntry] {
+	return func(yield func(Key, mapStateEntry) bool) {
+		iter := ms.trie.AncestorLongestPrefixFirstIterator(key.PrefixLength(), key)
+		for ok, lpmKey, idSet := iter.Next(); ok; ok, lpmKey, idSet = iter.Next() {
+			k := Key{LPMKey: lpmKey.Value()}
+
+			// Visit key with the same identity, if port/proto is different.
+			if !ms.forID(k.WithIdentity(key.Identity), idSet, yield) {
+				return
+			}
+			// Then visit key with zero identity if not already done above and one
+			// exists
+			if key.Identity != 0 && !ms.forID(k.WithIdentity(0), idSet, yield) {
+				return
+			}
+		}
+	}
+}
+
+// Lookup finds the policy verdict applicable to the given 'key' using the same precedence logic
+// between L3 and L4-only policies when both match the given 'key'.
+// To be used in testing in place of the bpf datapath when full integration testing is not desired.
+// Returns the closest matching covering policy entry and 'true' if found.
+// 'key' must not have a wildcard identity or port.
+func (ms *mapState) Lookup(key Key) (mapStateEntry, bool) {
```

Comment:
unfortunate that an exported function returns an unexported type. Can this return something simpler? (And maybe an unexported `lookup()` that returns `mapStateEntry`

Reply:
though I suppose it's somewhat moot, given that the receiver `mapState` is also unexported.

Reply:
Changed this to return the exported MapStateEntry member only. AFAIK this can be used from other packages if there is a function that returns `*mapState` to them, i.e., the receiver type need not be exported and it's exported functions can still be used from other packages.